src/io/flutter/perf/EditorPerfDecorations.java

```diff
@@ -0,0 +1,502 @@
+    }
+  }
+
+  @Override
+  public void clear() {
+    stats.clear();
+    removeHighlightersFromEditor();
+  }
+}
+
+class PerfGutterIconRenderer extends GutterIconRenderer {
+  static final AnimatedIcon RED_PROGRESS = new RedProgress();
+  static final AnimatedIcon NORMAL_PROGRESS = new AnimatedIcon.Grey();
+
+  private static final Icon EMPTY_ICON = new EmptyIcon(FlutterIcons.CustomInfo);
+
+  // Threshold for statistics to use red icons.
+  private static final int HIGH_LOAD_THRESHOLD = 100;
+
+  // Speed of the animation in radians per second.
+  private static final double ANIMATION_SPEED = 4.0;
+
+  private final RangeHighlighter highlighter;
+  private final TextRange range;
+  private final EditorPerfModel perfModelForFile;
+
+  // Tracked so we know when to notify that our icon has changed.
+  private Icon lastIcon;
+
+  PerfGutterIconRenderer(TextRange range,
+                         EditorPerfModel perfModelForFile,
+                         RangeHighlighter highlighter) {
+    this.highlighter = highlighter;
+    this.range = range;
+    this.perfModelForFile = perfModelForFile;
+    final TextAttributes textAttributes = highlighter.getTextAttributes();
+    assert textAttributes != null;
+    textAttributes.setEffectType(EffectType.ROUNDED_BOX);
+
+    updateUI(false);
+  }
+
+  public boolean isNavigateAction() {
+    return isActive();
+  }
+
+  private FlutterApp getApp() {
+    return perfModelForFile.getApp();
+  }
+
+  private int getCountPastSecond() {
+    return perfModelForFile.getStats().getCountPastSecond(range);
+  }
+
+  private boolean isActive() {
+    return perfModelForFile.isHoveredOverLineMarkerArea() || getCountPastSecond() > 0;
+  }
+
+  RangeHighlighter getHighlighter() {
+    return highlighter;
+  }
+
+  /**
+   * Returns the action executed when the icon is left-clicked.
+   *
+   * @return the action instance, or null if no action is required.
+   */
+  @Nullable
+  public AnAction getClickAction() {
+    return new AnAction() {
+      @Override
+      public void actionPerformed(AnActionEvent e) {
+        if (isActive()) {
+
+          final ToolWindowManagerEx toolWindowManager = ToolWindowManagerEx.getInstanceEx(getApp().getProject());
+          final ToolWindow flutterToolWindow = toolWindowManager.getToolWindow(FlutterView.TOOL_WINDOW_ID);
+          if (flutterToolWindow.isVisible()) {
+            showPerfViewMessage();
+            return;
+          }
+          flutterToolWindow.show(() -> showPerfViewMessage());
+        }
+      }
+    };
+  }
+
+  private void showPerfViewMessage() {
+    final FlutterView flutterView = ServiceManager.getService(getApp().getProject(), FlutterView.class);
+    final InspectorPerfTab inspectorPerfTab = flutterView.showPerfTab(getApp());
+    final StringBuilder sb = new StringBuilder("<html><body>");
+    final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("HH:mm:ss");
+    sb.append("<h2>Widget rebuild stats for ");
```

Comment:
Use bold instead of h2 for the title? I think we generally want to avoid mixing too many fonts in any one UI.

Reply:
Agreed. Done.